### ISSUE TITLE
helm still ignores kubeconfig

### ISSUE THREAD

**Issue by keyolk** [kubernetes/helm#2520]
Seems it related to this issue https://github.com/kubernetes/helm/issues/2379.
I use v2.4.2 helm, but it is not resolved.

See below
```
$ helm version
Client: &version.Version{SemVer:"v2.4.2", GitCommit:"82d8e9498d96535cc6787a6a9194a76161d29b4c", GitTreeState:"clean"}

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.4", GitCommit:"d6f433224538d4f9ca2f7ae19b252e6fcb66a3ae", GitTreeState:"clean", BuildDate:"2017-05-19T18:44:27Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.7", GitCommit:"8eb75a5810cba92ccad845ca360cf924f2385881", GitTreeState:"clean", BuildDate:"2017-04-27T09:42:05Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}

$ cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

$ env | grep KUBECONFIG
KUBECONFIG=/home1/irteam/.kube/config

$ cat $KUBECONFIG
---
apiVersion: v1
kind: Config
current-context: default
contexts:
- context:
    cluster: default
    user: admin
  name: default
clusters:
- cluster:
    server: https://apiserver.cluster.internal
    insecure-skip-tls-verify: true
  name: default
users:
- user:
    username: system
    password: anonymous  
  name: admin

$ helm init
$HELM_HOME has been configured at /home1/irteam/.helm.

Tiller (the helm server side component) has been installed into your Kubernetes Cluster.
Happy Helming!

$ kubectl logs tiller-deploy-3354596499-4kvgg -n kube-system
Cannot initialize Kubernetes connection: Get http://localhost:8080/api: dial tcp 127.0.0.1:8080: getsockopt: connection refused
```

Any Idea to solve this ?

**Comment by fibonacci1729** [comment]
What is the output of ```kubectl config current-context```? To me, this seems like your kubectl is set to the wrong context. The output of ```helm init``` indicates that tiller was installed into your cluster and presumably the one which you expected.

**Comment by keyolk** [comment]
Thanks  @fibonacci1729 
But it seems there is no error

Here my terminal outputs 
```
$ kubectl config current-context
default
$ kubectl config view
apiVersion: v1
clusters:
- cluster:
    insecure-skip-tls-verify: true
    server: https://apiserver.cluster.internal
  name: default
contexts:
- context:
    cluster: default
    user: admin
  name: default
current-context: default
kind: Config
preferences: {}
users:
- name: admin
  user:
    password: anonymous
    username: system
$ kubectl delete -n kube-system svc tiller-deploy
service "tiller-deploy" deleted
$ kubectl delete -n kube-system deploy tiller-deploy
deployment "tiller-deploy" deleted
$  helm init --debug --dry-run                                                                                     
apiVersion: extensions/v1beta1                                                                                                              
kind: Deployment                                                                                                                            
Spec:                                                                                                                                       
  MinReadySeconds: 0                                                                                                                        
  Paused: false                                                                                                                             
  ProgressDeadlineSeconds: null                                                                                                             
  Replicas: 1                                                                                                                               
  RevisionHistoryLimit: null                                                                                                                
  RollbackTo: null                                                                                                                          
  Selector: null                                                                                                                            
  Strategy:                                                                                                                                 
    RollingUpdate: null                                                                                                                     
    Type: ""                                                                                                                                
  Template:                                                                                                                                 
    Spec:                                                                                                                                   
      ActiveDeadlineSeconds: null                                                                                                           
      Affinity: null                                                                                                                        
      AutomountServiceAccountToken: null                                                                                                    
      Containers:                                                                                                                           
      - Args: null                                                                                                                          
        Command: null                                                                                                                       
        Env:                                                                                                                                
        - Name: TILLER_NAMESPACE                                                                                                            
          Value: kube-system                                                                                                                
          ValueFrom: null                                                                                                                   
        EnvFrom: null                                                                                                                       
        Image: gcr.io/kubernetes-helm/tiller:v2.4.2                                                                                         
        ImagePullPolicy: IfNotPresent                                                                                                       
        Lifecycle: null                                                                                                                     
        LivenessProbe:                                                                                                                      
          Exec: null                                                                                                                        
          FailureThreshold: 0                                                                                                               
          HTTPGet:                                                                                                                          
            HTTPHeaders: null                                                                                                               
            Host: ""                                                                                                                        
            Path: /liveness                                                                                                                 
            Port: 44135              
            Scheme: ""
          InitialDelaySeconds: 1
          PeriodSeconds: 0
          SuccessThreshold: 0
          TCPSocket: null
          TimeoutSeconds: 1
        Name: tiller
        Ports:
        - ContainerPort: 44134
          HostIP: ""
          HostPort: 0
          Name: tiller
          Protocol: ""
        ReadinessProbe:
          Exec: null
          FailureThreshold: 0
          HTTPGet:
            HTTPHeaders: null
            Host: ""
            Path: /readiness
            Port: 44135
            Scheme: ""
          InitialDelaySeconds: 1
          PeriodSeconds: 0
          SuccessThreshold: 0
          TCPSocket: null
          TimeoutSeconds: 1
        Resources:
          Limits: null
          Requests: null
        SecurityContext: null
        Stdin: false
        StdinOnce: false
        TTY: false
        TerminationMessagePath: ""
        TerminationMessagePolicy: ""                                                                                                        
        VolumeMounts: null                                                                                                                  
        WorkingDir: ""                                                                                                                      
      DNSPolicy: ""                                                                                                                         
      Hostname: ""                                                                                                                          
      ImagePullSecrets: null                                                                                                                
      InitContainers: null                                                                                                                  
      NodeName: ""                                                                                                                          
      NodeSelector: null                                                                                                                    
      RestartPolicy: ""                                                                                                                     
      SchedulerName: ""                                                                                                                     
      SecurityContext:                                                                                                                      
        FSGroup: null                                                                                                                       
        HostIPC: false                                                                                                                      
        HostNetwork: false                                                                                                                  
        HostPID: false                                                                                                                      
        RunAsNonRoot: null                                                                                                                  
        RunAsUser: null                                                                                                                     
        SELinuxOptions: null                                                                                                                
        SupplementalGroups: null                                                                                                            
      ServiceAccountName: ""
      Subdomain: ""
      TerminationGracePeriodSeconds: null
      Tolerations: null
      Volumes: null
    creationTimestamp: null
    labels:
      app: helm
      name: tiller
Status:
  AvailableReplicas: 0
  Conditions: null
  ObservedGeneration: 0
  ReadyReplicas: 0
  Replicas: 0
  UnavailableReplicas: 0
  UpdatedReplicas: 0
creationTimestamp: null
labels:
  app: helm
  name: tiller
name: tiller-deploy
namespace: kube-system
---
apiVersion: v1
kind: Service
Spec:
  ClusterIP: ""
  ExternalIPs: null
  ExternalName: ""
  LoadBalancerIP: ""
  LoadBalancerSourceRanges: null
  Ports:
  - Name: tiller
    NodePort: 0
    Port: 44134
    Protocol: ""
    TargetPort: tiller
  Selector:
    app: helm
    name: tiller
  SessionAffinity: ""
  Type: ClusterIP
Status:
  LoadBalancer:
    Ingress: null
creationTimestamp: null
labels:
  app: helm
  name: tiller
name: tiller-deploy
namespace: kube-system
...
Creating /home1/irteam/.helm 
Creating /home1/irteam/.helm/repository 
Creating /home1/irteam/.helm/repository/cache 
Creating /home1/irteam/.helm/repository/local 
Creating /home1/irteam/.helm/plugins 
Creating /home1/irteam/.helm/starters 
Creating /home1/irteam/.helm/repository/repositories.yaml 
$HELM_HOME has been configured at /home1/irteam/.helm.

Tiller (the helm server side component) has been installed into your Kubernetes Cluster.
Happy Helming!

$ kubectl get pod -n kube-system | grep tiller
tiller-deploy-3354596499-8qtsh           0/1       ContainerCreating   0          1m

$ kubectl -n kube-system logs tiller-deploy-3354596499-8qtsh
Cannot initialize Kubernetes connection: Get http://localhost:8080/api: dial tcp 127.0.0.1:8080: getsockopt: connection refused
```

**Comment by technosophos** [comment]
If `helm init` correctly intalled Tiller, then your $KUBECONFIG is not the problem. `helm` reads it in order to install Tiller, and Tiller was installed.

Instead, it seems that inside of your cluster, Tiller cannot connect to the API server. You might want to `kubectl exec -it` onto the Tiller pod and check out whether the Kubernetes secrets were correctly mounted to the pod. Usually, running `mount` will show you that.

**Comment by keyolk** [comment]
@technosophos seems it is not installed correctly.
the tiller alwasy dial to 127.0.0.1:8080 which is not my api server.

**Comment by thomasjungblut** [comment]
any fixes for it? I'm running into the same issue.  $KUBECONFIG is basically not honored when running helm.

**Comment by bacongobbler** [comment]
@thomasjungblut what is the error you're seeing? Can you post more information about what version, the status of tiller (`kubectl get pod -n kube-system | grep tiller`) etc?

**Comment by bacongobbler** [comment]
As shown in this demonstration, $KUBECONFIG works quite well with v2.6.0.

```
><> #running helm with no kubeconfig
><> docker run -v $PWD:/tmp busybox /tmp/linux-amd64/helm list
Error: Get http://localhost:8080/api/v1/namespaces/kube-system/pods?labelSelector=app%3Dhelm%2Cname%3Dtiller: dial tcp 127.0.0.1:8080: getsockopt: connection refused

><> # running helm with mounted kubeconfig at $HOME/.kube/config
><> docker run -v $PWD:/tmp -v ~/.kube/config:/root/.kube/config busybox /tmp/linux-amd64/helm list
Error: could not get Kubernetes config for context "": invalid configuration: [unable to read client-cert /Users/bacongobbler/.minikube/apiserver.crt for minikube due to open /Users/bacongobbler/.minikube/apiserver.crt: no such file or directory, unable to read client-key /Users/bacongobbler/.minikube/apiserver.key for minikube due to open /Users/bacongobbler/.minikube/apiserver.key: no such file or directory, unable to read certificate-authority /Users/bacongobbler/.minikube/ca.crt for minikube due to open /Users/bacongobbler/.minikube/ca.crt: no such file or directory]

><> # helm with kubeconfig mounted at $HOME/kubeconfig, no $KUBECONFIG envvar present
><> docker run -v $PWD:/tmp -v ~/.kube/config:/root/kubeconfig busybox /tmp/linux-amd64/helm list
Error: Get http://localhost:8080/api/v1/namespaces/kube-system/pods?labelSelector=app%3Dhelm%2Cname%3Dtiller: dial tcp 127.0.0.1:8080: getsockopt: connection refused

><> # helm with kubeconfig mounted at $HOME/kubeconfig, setting $KUBECONFIG to point at that path
><> docker run -v $PWD:/tmp -v ~/.kube/config:/root/kubeconfig -e KUBECONFIG=/root/kubeconfig busybox /tmp/linux-amd64/helm list
Error: could not get Kubernetes config for context "": invalid configuration: [unable to read client-cert /Users/bacongobbler/.minikube/apiserver.crt for minikube due to open /Users/bacongobbler/.minikube/apiserver.crt: no such file or directory, unable to read client-key /Users/bacongobbler/.minikube/apiserver.key for minikube due to open /Users/bacongobbler/.minikube/apiserver.key: no such file or directory, unable to read certificate-authority /Users/bacongobbler/.minikube/ca.crt for minikube due to open /Users/bacongobbler/.minikube/ca.crt: no such file or directory]
```

**Comment by bacongobbler** [comment]
closing as a duplicate of https://github.com/kubernetes/helm/issues/2619 but please re-open if necessary.

**Comment by vishveshv** [comment]
Any fixes for this? I am also running into same issue:

$> kb get pod -n kube-system | grep tiller
tiller-deploy-3718853040-hrw8t   1/1       Running   1          7d

$>  hm install <helm-chart> -name=<chartName>
Error: Get http://localhost:8080/version: dial tcp 127.0.0.1:8080: getsockopt: connection refused

It seems tiller server isn't not able to connect with api server but why?

**Comment by mjbright** [comment]
I'm also seeing this problem trying to use helm with minikube.

Running minikube 0.26.1 and helm v2.9.0-rc4 on macos.

Nothing else to add that the original poster hasn't already detailed.

Here's my .kube/config showing that minikube is the default/only config:

`> cat ~/.kube/config 
apiVersion: v1
clusters:
- cluster:
    certificate-authority: /Users/mjb/.minikube/ca.crt
    server: https://192.168.99.100:8443
  name: minikube
contexts:
- context:
    cluster: minikube
    user: minikube
  name: minikube
current-context: minikube
kind: Config
preferences: {}
users:
- name: minikube
  user:
    client-certificate: /Users/mjb/.minikube/client.crt
    client-key: /Users/mjb/.minikube/client.key`

**Comment by mjbright** [comment]
Here's the log of the tiller pod:

> > kubectl -n kube-system logs tiller-deploy-75d848bb9-fgbxw
> [main] 2018/04/20 14:37:00 Starting Tiller v2.9.0-rc4 (tls=false)
> [main] 2018/04/20 14:37:00 GRPC listening on :44134
> [main] 2018/04/20 14:37:00 Probes listening on :44135
> [main] 2018/04/20 14:37:00 Storage driver is ConfigMap
> [main] 2018/04/20 14:37:00 Max history per release is 0
> [tiller] 2018/04/20 14:37:45 preparing install for broker
> [storage] 2018/04/20 14:37:45 getting release history for "broker"
> [storage/driver] 2018/04/20 14:37:45 query: failed to query with labels: Get http://localhost:8080/api/v1/namespaces/kube-system/configmaps?labelSelector=NAME%3Dbroker%2COWNER%3DTILLER: dial tcp 127.0.0.1:8080: connect: connection refused
> [tiller] 2018/04/20 14:37:45 failed install prepare step: Get http://localhost:8080/version: dial tcp 127.0.0.1:8080: connect: connection refused

**Comment by bacongobbler** [comment]
@mjbright that might actually be unrelated here. It seems like your API server is being unresponsive from within the cluster. I think there's another issue open with this issue. Try searching for open issues with "connection refused" on the issue queue.

**Comment by mjbright** [comment]
Thanks for that @bacongobbler, if I hit the problem again I'll search for "connection refused" issues as you suggest.

In  the meantime I moved back to helm 2.7.0 (because I was trying to reproduce a [blog article](https://medium.com/@JakubBorys/run-nameko-services-on-kubernetes-c4db3a863f05) which used that version) and this fixed my problem.

**Comment by delcon** [comment]
I am having the same issue here:
`PS D:\Users\leove> helm install stable/voyager --name my-release --set cloudProvider=$provider
Error: Get http://localhost:8080/version: dial tcp [::1]:8080: connect: connection refused`

**Comment by garystafford** [comment]
I moved from the latest, `2.9.0`, back to `2.8.2`, and it worked. The problem seems to be with `2.9.0`.

```text
helm version
Client: &version.Version{SemVer:"v2.9.0", GitCommit:"f6025bb9ee7daf9fee0026541c90a6f557a3e0bc", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.0", GitCommit:"f6025bb9ee7daf9fee0026541c90a6f557a3e0bc", GitTreeState:"clean"}

helm install --name my-release stable/spinnaker
Error: Get http://localhost:8080/version: dial tcp [::1]:8080: connect: connection refused
```
```text
./helm version
Client: &version.Version{SemVer:"v2.8.2", GitCommit:"a80231648a1473929271764b920a8e346f6de844", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.2", GitCommit:"a80231648a1473929271764b920a8e346f6de844", GitTreeState:"clean"}
```

Got passed that error...

**Comment by arun-gupta** [comment]
+1

```
$ helm version
Client: &version.Version{SemVer:"v2.9.0", GitCommit:"f6025bb9ee7daf9fee0026541c90a6f557a3e0bc", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.0", GitCommit:"f6025bb9ee7daf9fee0026541c90a6f557a3e0bc", GitTreeState:"clean"}
```

Can't even delete Helm:

```
$ helm reset
Error: Get http://localhost:8080/api/v1/namespaces/kube-system/configmaps?labelSelector=OWNER%!D(MISSING)TILLER: dial tcp [::1]:8080: connect: connection refused
```

Had to delete as:

```
$ kubectl delete deployment tiller-deploy --namespace kube-system
deployment.extensions "tiller-deploy" deleted
```

Switched to 2.8.1:

```
$ brew switch kubernetes-helm 2.8.1
Cleaning /usr/local/Cellar/kubernetes-helm/2.6.2
Cleaning /usr/local/Cellar/kubernetes-helm/2.8.1
Cleaning /usr/local/Cellar/kubernetes-helm/2.9.0
46 links created for /usr/local/Cellar/kubernetes-helm/2.8.1
```

Now encountering https://github.com/kubernetes/helm/issues/3130

**Comment by maxyor6040** [comment]
Don't forget to `minikube delete` and `minikube start` after downgrading. Then `helm init` again.

**Comment by disha94** [comment]
Facing the same issue, what is the fix for the same?

**Comment by maxyor6040** [comment]
@disha94 

I'm on macbook pro 2017 with high sierra.
This solved the issue for me:

1. Downgrade helm to 2.8.2
2. `minikube delete`
3. `minikube start`
4. `helm init --force-upgrade`

**Comment by bacongobbler** [comment]
Also see #3985 for a (few) workarounds. I'm just coming off vacation today and will see what I can do about getting a v2.9.1 out today.

**Comment by disha94** [comment]
@maxyor6040 Thanks a lot for your help. But I am facing the issue while deploying on a Kubernetes cluster and not in minikube. But looks like upgrade might help.
@bacongobbler , is this solution specific to Azure?? As in my case cloud provider is AWS.

**Comment by bacongobbler** [comment]
it's not specific to Azure. I'l update the title to reflect that. Thanks for pointing that out. :)

**Comment by moritzheiber** [comment]
Would it be good to re-open this issue? It appears to be happening in a reproducible way.

**Comment by bacongobbler** [comment]
We're pretty sure this has been fixed in 2.9.1 with e clean install. Can you confirm?

**Comment by moritzheiber** [comment]
No, unfortunately I cannot confirm this is resolved with 2.9.1. On a new cluster, provisioned with kops, this remains an issue.

**Comment by bacongobbler** [comment]
@moritzheiber kops bootstraps a cluster with role based access controls enabled. You'll need to follow https://docs.helm.sh/using_helm/#role-based-access-control to configure Tiller with the right roles and rolebindings to work with kops clusters.

**Comment by nikhilbandi** [comment]
I encountered this issue initially when trying to install a chart using helm. 
After a lot of debugging, I realised there was an issue with one of the container images, due to which the container gets killed and restarted. However, kubectl tries to communicate with the container and is not able to do so. That's why it gives this error message. When I resolved the issue with the container image, the 'helm install' worked fine and containers are running smoothly.
